### PR TITLE
Fix Falcon remainder validation and u64::rotr overflow handling

### PR DESCRIPTION
**Pull request by huitseeker**
This PR fixes two correctness bugs.

In Falcon `mod_12289`, the advice-supplied remainder is now validated as a `u32` before any u32 operation uses it. This blocks forged non-`u32` remainders from producing the wrong result. Tests cover the forged-remainder path.

In `u64::rotr`, the field arithmetic path is replaced with `u32widening_mul` and `u32widening_madd`. This fixes bad outputs for rotations by `0` and `32` when the input is at or above the Goldilocks modulus. Tests cover both cases.

The `rotr` docs now match in the MASM source, generated core docs, and user docs.

### CODE REVIEW

File: crates/lib/core/asm/crypto/dsa/falcon512_poseidon2.masm [hunk 1]
```diff
@@ -55,6 +55,9 @@ pub proc mod_12289
     # => [res_hi, res_lo, a_hi, a_lo, ...]
 
     adv_push.1
+    # Advice values are untrusted. U32SUB/U32ADD only constrain their output limbs in the AIR,
+    # so we must validate the remainder before feeding it into u32 arithmetic.
+    u32assert
     dup
```
Comment (bobbinth) [inline crates/lib/core/asm/crypto/dsa/falcon512_poseidon2.masm]: `u32assert` takes 2 VM cycles because it is a macro instruction that uses `u32assert2` under the hood. But here, we could just use `u32assert2` doing something like:

```
adv_push.1
dup
u32assert2
```
This way, we'll introduce only 1 extra cycle.

File: crates/lib/core/asm/math/u64.masm [hunk 2]
```diff
@@ -826,115 +826,130 @@ end
...
 
     # --- STEP 5: Carry the overflow to the low bits ---
     movup.2
-    # [lo1, hi2, lo2, flag, ...]
+    # [lo1, hi2, lo2, borrow, shift_amt == 0, ...]
 
     add
-    # [lo1 + hi2, lo2, flag, ...]
+    # [lo1 + hi2, lo2, borrow, shift_amt == 0, ...]
     # c_lo = lo1 + hi2 (overflow wraps to low bits)
```
Comment (Al-Kindi-0) [inline crates/lib/core/asm/math/u64.masm:938]: worth adding a comment about the safety of using `add` on u32 here

Comment (Al-Kindi-0) [inline crates/lib/core/asm/math/u64.masm:938]: worth adding a comment about the safety of using `add` on u32 here

File: crates/lib/core/tests/math/u64_mod.rs [hunk 3]
```diff
@@ -1203,6 +1204,36 @@ fn unchecked_rotr() {
     build_test!(source, &stack![b as u64, a0, a1, 5]).expect_stack(&[c0, c1, 5]);
 }
 
+#[test]
+fn unchecked_rotr_large_value_by_0() {
```
Comment (Al-Kindi-0) [inline crates/lib/core/tests/math/u64_mod.rs:1208]: Should we also add tests, especially at the edge points, for `rotl` as well ?

File: crates/lib/core/asm/math/u64.masm (hunk 2 above)
Comment (Al-Kindi-0) [reply crates/lib/core/asm/math/u64.masm:938]: Seems worth also add a similar comment about the use of `add` a bit up for `shr` procedure.
Maybe good practice for the future to always argue safety of using field ops on non-field elements

File: air/src/constraints/stack/stack_arith/tests.rs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+use super::enforce_main;
+use crate::{
+    MainTraceRow,
+    constraints::op_flags::{ExprDecoderAccess, OpFlags, generate_test_row},
+    trace::{
+        AUX_TRACE_RAND_CHALLENGES, AUX_TRACE_WIDTH, TRACE_WIDTH, decoder::USER_OP_HELPERS_OFFSET,
+    },
+};
+
+struct ConstraintEvalBuilder {
```
Comment (adr1anh) [inline air/src/constraints/stack/stack_arith/tests.rs:22]: Looks like a nice reusable abstraction, I'll see how it could fit with #3004
